presto-password-authenticators/src/main/java/io/prestosql/plugin/password/ldap/LdapAuthenticator.java

```diff
@@ -171,9 +173,88 @@ private void checkForGroupMembership(String user, DirContext context)
         }
     }
 
-    private static String replaceUser(String pattern, String user)
+    @VisibleForTesting
+    static String replaceUserInSearch(String pattern, String user)
     {
-        return pattern.replaceAll("\\$\\{USER}", user);
+        return pattern.replaceAll("\\$\\{USER}", Matcher.quoteReplacement(encodeForSearch(user)));
+    }
+
+    @VisibleForTesting
+    static String replaceUserInBind(String pattern, String user)
+    {
+        return pattern.replaceAll("\\$\\{USER}", Matcher.quoteReplacement(encodeForDN(user)));
+    }
+
+    // Based on https://www.owasp.org/index.php/Preventing_LDAP_Injection_in_Java
+    @VisibleForTesting
+    static String encodeForSearch(String input)
+    {
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < input.length(); i++) {
+            char c = input.charAt(i);
+            switch (c) {
+                case '\\':
+                    sb.append("\\5c");
+                    break;
+                case '*':
+                    sb.append("\\2a");
+                    break;
+                case '(':
+                    sb.append("\\28");
+                    break;
+                case ')':
+                    sb.append("\\29");
+                    break;
+                case '\u0000':
+                    sb.append("\\00");
+                    break;
+                default:
+                    sb.append(c);
+            }
+        }
+        return sb.toString();
+    }
+
+    // Based on https://www.owasp.org/index.php/Preventing_LDAP_Injection_in_Java
+    @VisibleForTesting
+    static String encodeForDN(String input)
+    {
+        StringBuilder sb = new StringBuilder();
+        if ((input.length() > 0) && ((input.charAt(0) == ' ') || (input.charAt(0) == '#'))) {
```

Comment:
Use `isEmpty`

Reply:
I just found an existing method javax.naming.ldap.Rdn.escapeValue() which works for DN attributes. I didn't find anything similar for escaping attributes in search queries.